core/src/main/java/io/questdb/cairo/CairoEngine.java

```diff
@@ -588,24 +607,80 @@ public void unlock(
             CairoSecurityContext securityContext,
             CharSequence tableName,
             @Nullable TableWriter writer,
-            boolean newTable
+            boolean newTable,
+            Path path
     ) {
+        securityContext.checkWritePermission();
         validNameOrThrow(tableName);
         readerPool.unlock(tableName);
+        unlockReaders(tableName, path);
         writerPool.unlock(tableName, writer, newTable);
         metadataPool.unlock(tableName);
         LOG.info().$("unlocked [table=`").utf8(tableName).$("`]").$();
     }
 
-    public void unlockReaders(CharSequence tableName) {
+    public void unlockReaderPool(CharSequence tableName) {
         validNameOrThrow(tableName);
         readerPool.unlock(tableName);
     }
 
-    public void unlockWriter(CairoSecurityContext securityContext, CharSequence tableName) {
-        securityContext.checkWritePermission();
-        validNameOrThrow(tableName);
-        writerPool.unlock(tableName);
+    private boolean lockReaders(CharSequence tableName, Path path) {
+        // this method must be invoked only when thread has lock on writer pool
+        path.of(configuration.getRoot()).concat(tableName);
+        int plen = path.length();
+
+        final FilesFacade ff = configuration.getFilesFacade();
+        try (TxnScoreboard txnScoreboard = new TxnScoreboard(configuration)) {
+            if (txnScoreboard.ofRW(path, true) != null) {
+                long txn = readMaxTxn(ff, path.trimTo(plen));
+                if (txn == -2) {
+                    // message would have already been logged
+                    return false;
+                }
+
+                // make sure we can get a slot
+                if (txnScoreboard.acquireTxn(txn)) {
+                    txnScoreboard.releaseTxn(txn);
+
+                    // lock the current txn if scoreboard is at min
+                    if (txnScoreboard.getMin() == txn && txnScoreboard.releaseTxn(txn) == -1) {
```

Comment:
Also in case of failed attempt to lock, readers cannot be created on another engine until `unlockReaders` is called here. So the lock failed but prevents other engine to read.